File: src/lang.ts
/**
 * 扩展名 → shiki 语言 ID 映射。
 * 未知扩展名返回 null（→ 无色纯文本，用作高亮断言对照组）。
 */
const EXT_MAP: Record<string, string> = {
  ts: "typescript",
  tsx: "tsx",
  mts: "typescript",
  cts: "typescript",
  js: "javascript",
  jsx: "jsx",
  mjs: "javascript",
  cjs: "javascript",
  py: "python",
  pyi: "python",
  rs: "rust",
  go: "go",
  java: "java",
  c: "c",
  h: "c",
  cpp: "cpp",
  cc: "cpp",
  cxx: "cpp",
  hpp: "cpp",
  hxx: "cpp",
  cs: "csharp",
  rb: "ruby",
  php: "php",
  swift: "swift",
  kt: "kotlin",
  dart: "dart",
  scala: "scala",
  sh: "bash",
  bash: "bash",
  zsh: "bash",
  fish: "bash",
  json: "json",
  jsonc: "json",
  yaml: "yaml",
  yml: "yaml",
  toml: "toml",
  ini: "ini",
  html: "html",
  htm: "html",
  xml: "xml",
  css: "css",
  scss: "scss",
  less: "less",
  vue: "vue",
  svelte: "svelte",
  sql: "sql",
  graphql: "graphql",
  gql: "graphql",
  md: "markdown",
  markdown: "markdown",
  dockerfile: "dockerfile",
  makefile: "makefile",
  lua: "lua",
  r: "r",
  pl: "perl",
  pm: "perl",
  diff: "diff",
  patch: "diff",
  bat: "bat",
  ps1: "powershell",
  ps: "powershell",
};

/** 特殊文件名（无扩展名但有意义的语言）。 */
const NAME_MAP: Record<string, string> = {
  dockerfile: "dockerfile",
  makefile: "makefile",
  justfile: "makefile",
  gemfile: "ruby",
  rakefile: "ruby",
};

export function detectLang(fileName: string): string | null {
  const base = fileName.split("/").pop() ?? fileName;
  const lower = base.toLowerCase();
  if (NAME_MAP[lower]) return NAME_MAP[lower];
  const dot = lower.lastIndexOf(".");
  if (dot < 0) return null;
  const ext = lower.slice(dot + 1);
  return EXT_MAP[ext] ?? null;
}

/** 是否为 markdown 扩展名（走 TVirtualMarkdown 渲染）。 */
export function isMarkdownExt(fileName: string): boolean {
  const base = fileName.split("/").pop() ?? fileName;
  const lower = base.toLowerCase();
  const dot = lower.lastIndexOf(".");
  if (dot < 0) return false;
  const ext = lower.slice(dot + 1);
  return ext === "md" || ext === "markdown";
}

File: rs/src/lang.rs
//! 扩展名 → 语言 ID / 模式映射。
//! .md/.markdown → Markdown 模式；.mmd/.mermaid → Mermaid 模式；其它 → Code 模式。
//!
//! 语法集为 two-face 全量 Sublime 语法(DECISIONS D4),覆盖 TS/Vue/Svelte/TOML/
//! INI/GraphQL/Dockerfile/PowerShell/SCSS/Less/Swift/Kotlin/Dart 等;
//! 个别缺失语言由 Highlighter::find_syntax 的回退链兜底。

/// 预览模式。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Mode {
    Markdown,
    Code,
    Mermaid,
}

/// 扩展名 → syntect token(先按扩展名查,再按 token 查,见 Highlighter::find_syntax)。
fn ext_to_syntax_token(ext: &str) -> Option<&'static str> {
    Some(match ext {
        "ts" | "mts" | "cts" => "ts",
        "tsx" => "tsx",
        "js" | "mjs" | "cjs" => "js",
        "jsx" => "jsx",
        "py" | "pyi" => "py",
        "rs" => "rs",
        "go" => "go",
        "java" => "java",
        "c" | "h" => "c",
        "cpp" | "cc" | "cxx" | "hpp" | "hxx" => "cpp",
        "cs" => "cs",
        "rb" => "rb",
        "php" => "php",
        "swift" => "swift",
        "kt" | "kts" => "kotlin",
        "dart" => "dart",
        "scala" => "scala",
        "sh" | "bash" | "zsh" | "fish" => "sh",
        "json" | "jsonc" => "json",
        "yaml" | "yml" => "yaml",
        "toml" => "toml",
        "ini" | "cfg" | "conf" => "ini",
        "html" | "htm" => "html",
        "xml" => "xml",
        "css" => "css",
        "scss" => "scss",
        "less" => "less",
        "vue" => "vue",
        "svelte" => "svelte",
        "sql" => "sql",
        "graphql" | "gql" => "graphql",
        "md" | "markdown" => "md",
        "lua" => "lua",
        "r" => "R",
        "pl" | "pm" => "pl",
        "diff" | "patch" => "diff",
        "bat" | "cmd" => "bat",
        "ps1" | "ps" | "psm1" => "ps1",
        _ => return None,
    })
}

/// 特殊文件名（无扩展名但有意义的语言）→ syntect token。
fn name_to_syntax_token(name: &str) -> Option<&'static str> {
    Some(match name {
        "makefile" | "justfile" => "makefile",
        "gemfile" | "rakefile" => "rb",
        "dockerfile" => "dockerfile",
        _ => return None,
    })
}

/// 返回文件名对应的 syntect 扩展名 token（用于代码高亮查找）。
/// 仅 Code 模式有意义。
pub fn detect_syntax_token(file_name: &str) -> Option<&'static str> {
    let base = file_name.rsplit('/').next().unwrap_or(file_name);
    let lower = base.to_lowercase();
    if let Some(t) = name_to_syntax_token(&lower) {
        return Some(t);
    }
    match lower.rfind('.') {
        Some(dot) => ext_to_syntax_token(&lower[dot + 1..]),
        None => None,
    }
}

/// 是否为 markdown 扩展名（走 termimad 渲染）。
pub fn is_markdown_ext(file_name: &str) -> bool {
    let base = file_name.rsplit('/').next().unwrap_or(file_name);
    let lower = base.to_lowercase();
    match lower.rfind('.') {
        Some(dot) => {
            let ext = &lower[dot + 1..];
            ext == "md" || ext == "markdown"
        }
        None => false,
    }
}

/// 是否为 mermaid 扩展名。
pub fn is_mermaid_ext(file_name: &str) -> bool {
    let base = file_name.rsplit('/').next().unwrap_or(file_name);
    let lower = base.to_lowercase();
    match lower.rfind('.') {
        Some(dot) => {
            let ext = &lower[dot + 1..];
            ext == "mmd" || ext == "mermaid"
        }
        None => false,
    }
}

/// 判定预览模式 + 语法 token。
pub fn detect_mode_lang(file_name: &str) -> (Mode, Option<&'static str>) {
    if is_markdown_ext(file_name) {
        return (Mode::Markdown, None);
    }
    if is_mermaid_ext(file_name) {
        return (Mode::Mermaid, None);
    }
    (Mode::Code, detect_syntax_token(file_name))
}
